File: tests/org/elixir_lang/reference/callable/Issue463Test.java
package org.elixir_lang.reference.callable;

import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiPolyVariantReference;
import com.intellij.psi.PsiReference;
import com.intellij.psi.ResolveResult;
import com.intellij.testFramework.fixtures.BasePlatformTestCase;
import com.intellij.util.xml.Resolve;
import org.elixir_lang.psi.CallDefinitionClause;
import org.elixir_lang.psi.call.Call;

public class Issue463Test extends BasePlatformTestCase {
    /*
     * Tests
     */

    public void testDirectModuleQualifier() {
        myFixture.configureByFiles("direct_module_qualifier.ex", "referenced.ex");
        PsiElement elementAtCaret = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset());

        assertNotNull(elementAtCaret);

        PsiElement grandParent = elementAtCaret.getParent().getParent();

        assertNotNull(grandParent);
        assertInstanceOf(grandParent, Call.class);

        PsiReference reference = grandParent.getReference();

        assertNotNull(reference);

        PsiElement resolved = reference.resolve();

        assertNotNull(resolved);

        assertInstanceOf(resolved, Call.class);
        assertEquals(
                "def changeset(params) do\n" +
                "    %__MODULE__{}\n" +
                "    |> cast(params, ~w(name))\n" +
                "    |> validate_required(:name)\n" +
                "  end",
                resolved.getText()
        );

        assertInstanceOf(resolved, Call.class);

        assertTrue(CallDefinitionClause.is((Call) resolved));
    }

    public void testAliasedModuleQualifier() {
        myFixture.configureByFiles("aliased_module_qualifier.ex", "referenced.ex");
        PsiElement elementAtCaret = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset());

        assertNotNull(elementAtCaret);

        PsiElement grandParent = elementAtCaret.getParent().getParent();

        assertNotNull(grandParent);
        assertInstanceOf(grandParent, Call.class);

        PsiReference reference = grandParent.getReference();

        assertNotNull(reference);

        PsiElement resolved = reference.resolve();

        assertNotNull(resolved);

        assertInstanceOf(resolved, Call.class);
        assertEquals(
                "def changeset(params) do\n" +
                        "    %__MODULE__{}\n" +
                        "    |> cast(params, ~w(name))\n" +
                        "    |> validate_required(:name)\n" +
                        "  end",
                resolved.getText()
        );

        assertInstanceOf(resolved, Call.class);

        assertTrue(CallDefinitionClause.is((Call) resolved));
    }

    public void testDoubleAliasesModuleQualifier() {
        myFixture.configureByFiles("double_aliased_module_qualifier.ex", "referenced.ex");
        PsiElement elementAtCaret = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset());

        assertNotNull(elementAtCaret);

        PsiElement grandParent = elementAtCaret.getParent().getParent();

        assertNotNull(grandParent);
        assertInstanceOf(grandParent, Call.class);

        PsiReference reference = grandParent.getReference();

        assertNotNull(reference);

        PsiElement resolved = reference.resolve();

        assertNotNull(resolved);
        assertEquals(resolved.getText(), "def changeset(params) do\n" +
                "    %__MODULE__{}\n" +
                "    |> cast(params, ~w(name))\n" +
                "    |> validate_required(:name)\n" +
                "  end");
    }

    public void testMapAccessQualifier() {
        myFixture.configureByFiles("map_access_qualifier.ex", "referenced.ex");
        PsiElement elementAtCaret = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset());

        assertNotNull(elementAtCaret);

        PsiElement grandParent = elementAtCaret.getParent().getParent();

        assertNotNull(grandParent);
        assertInstanceOf(grandParent, Call.class);

        PsiReference reference = grandParent.getReference();

        assertNotNull(reference);

        PsiElement resolved = reference.resolve();

        assertNull(resolved);
    }

    public void testUnresolvedAliasQualifier() {
        myFixture.configureByFiles("unresolved_alias_qualifier.ex", "referenced.ex");
        PsiElement elementAtCaret = myFixture
                .getFile()
                .findElementAt(myFixture.getCaretOffset());

        assertNotNull(elementAtCaret);

        PsiElement grandParent = elementAtCaret.getParent().getParent();

        assertNotNull(grandParent);
        assertInstanceOf(grandParent, Call.class);

        PsiReference reference = grandParent.getReference();

        assertNotNull(reference);

        assertInstanceOf(reference, PsiPolyVariantReference.class);

        PsiPolyVariantReference psiPolyVariantReference = (PsiPolyVariantReference) reference;

        ResolveResult[] resolveResults = psiPolyVariantReference.multiResolve(false);

        assertEquals(1, resolveResults.length);

        ResolveResult firstResolveResult = resolveResults[0];

        assertFalse(firstResolveResult.isValidResult());

        PsiElement firstElement = firstResolveResult.getElement();

        assertNotNull(firstElement);

        assertEquals(firstElement.getText(), "def changeset(params) do\n" +
                "    %__MODULE__{}\n" +
                "    |> cast(params, ~w(name))\n" +
                "    |> validate_required(:name)\n" +
                "  end");
    }

    /*
     * Protected Instance Methods
     */

    @Override
    protected String getTestDataPath() {
        return "testData/org/elixir_lang/reference/callable/issue_463";
    }
}

File: tests/org/elixir_lang/reference/callable/Issue463Test.kt
package org.elixir_lang.reference.callable

import org.elixir_lang.psi.CallDefinitionClause.`is`
import com.intellij.psi.PsiPolyVariantReference
import org.elixir_lang.PlatformTestCase
import org.elixir_lang.psi.call.Call

class Issue463Test : PlatformTestCase() {
    fun testDirectModuleQualifier() {
        myFixture.configureByFiles("direct_module_qualifier.ex", "referenced.ex")
        val elementAtCaret = myFixture
                .file
                .findElementAt(myFixture.caretOffset)
        assertNotNull(elementAtCaret)

        val grandParent = elementAtCaret!!.parent.parent
        assertNotNull(grandParent)
        assertInstanceOf(grandParent, Call::class.java)

        val reference = grandParent.reference
        assertNotNull(reference)

        val resolved = reference!!.resolve()
        assertNotNull(resolved)
        assertInstanceOf(resolved, Call::class.java)
        assertEquals(
                """def changeset(params) do
    %__MODULE__{}
    |> cast(params, ~w(name))
    |> validate_required(:name)
  end""",
                resolved!!.text
        )
        assertInstanceOf(resolved, Call::class.java)
        assertTrue(`is`((resolved as Call?)!!))
    }

    fun testAliasedModuleQualifier() {
        myFixture.configureByFiles("aliased_module_qualifier.ex", "referenced.ex")
        val elementAtCaret = myFixture
                .file
                .findElementAt(myFixture.caretOffset)
        assertNotNull(elementAtCaret)

        val grandParent = elementAtCaret!!.parent.parent
        assertNotNull(grandParent)
        assertInstanceOf(grandParent, Call::class.java)

        val reference = grandParent.reference
        assertNotNull(reference)

        val resolved = reference!!.resolve()
        assertNotNull(resolved)
        assertInstanceOf(resolved, Call::class.java)
        assertEquals(
                """def changeset(params) do
    %__MODULE__{}
    |> cast(params, ~w(name))
    |> validate_required(:name)
  end""",
                resolved!!.text
        )
        assertInstanceOf(resolved, Call::class.java)
        assertTrue(`is`((resolved as Call?)!!))
    }

    fun testDoubleAliasesModuleQualifier() {
        myFixture.configureByFiles("double_aliased_module_qualifier.ex", "referenced.ex")
        val elementAtCaret = myFixture
                .file
                .findElementAt(myFixture.caretOffset)
        assertNotNull(elementAtCaret)

        val grandParent = elementAtCaret!!.parent.parent
        assertNotNull(grandParent)
        assertInstanceOf(grandParent, Call::class.java)

        val reference = grandParent.reference
        assertNotNull(reference)

        val resolved = reference!!.resolve()
        assertNotNull(resolved)
        assertEquals(resolved!!.text, """def changeset(params) do
    %__MODULE__{}
    |> cast(params, ~w(name))
    |> validate_required(:name)
  end""")
    }

    fun testMapAccessQualifier() {
        myFixture.configureByFiles("map_access_qualifier.ex", "referenced.ex")
        val elementAtCaret = myFixture
                .file
                .findElementAt(myFixture.caretOffset)
        assertNotNull(elementAtCaret)

        val grandParent = elementAtCaret!!.parent.parent
        assertNotNull(grandParent)
        assertInstanceOf(grandParent, Call::class.java)

        val reference = grandParent.reference
        assertNotNull(reference)

        val resolved = reference!!.resolve()
        assertNull(resolved)
    }

    fun testUnresolvedAliasQualifier() {
        myFixture.configureByFiles("unresolved_alias_qualifier.ex", "referenced.ex")
        val elementAtCaret = myFixture
                .file
                .findElementAt(myFixture.caretOffset)
        assertNotNull(elementAtCaret)

        val grandParent = elementAtCaret!!.parent.parent
        assertNotNull(grandParent)
        assertInstanceOf(grandParent, Call::class.java)

        val reference = grandParent.reference
        assertNotNull(reference)
        assertInstanceOf(reference, PsiPolyVariantReference::class.java)
        val psiPolyVariantReference = reference as PsiPolyVariantReference

        // An unresolvable qualifier (no alias, no module definition) should produce
        // zero results — not a global search across every module in the project.
        // The standard unresolved-reference highlight will inform the user.
        val resolveResults = psiPolyVariantReference.multiResolve(false)
        assertEquals(0, resolveResults.size)
    }

    override fun getTestDataPath(): String = "testData/org/elixir_lang/reference/callable/issue_463"
}
